File: src/web/magmac/app/stage/lexer/RuleLexer.ts
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { CompileResultCollector } from "../../../../magmac/app/compile/error/CompileResultCollector";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { UnitSetCollector } from "../../../../magmac/app/io/sources/UnitSetCollector";
import { Unit } from "../../../../magmac/app/stage/unit/Unit";
import { UnitSet } from "../../../../magmac/app/stage/unit/UnitSet";
export class RuleLexer {
	 RuleLexer( rootRule : Rule) : public {this.rootRule=rootRule;;}
	private foldEntry( unit : Unit<String>) : CompileResult<Unit<Node>> {System.out.println( unit.display( ));return unit.mapValue( this.rootRule.lex);;}
	public apply( initial : UnitSet<String>) : CompileResult<UnitSet<Node>> {return initial.iter( ).map( this.foldEntry).collect( new CompileResultCollector<>( new UnitSetCollector<>( )));;}
}

File: packages/compiler-java/src/main/java/magmac/app/stage/lexer/RuleLexer.java
package magmac.app.stage.lexer;

import magmac.app.compile.error.CompileResult;
import magmac.app.compile.error.CompileResultCollector;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.Rule;
import magmac.app.io.sources.UnitSetCollector;
import magmac.app.stage.unit.Unit;
import magmac.app.stage.unit.UnitSet;

public class RuleLexer implements Lexer {
    private final Rule rootRule;

    public RuleLexer(Rule rootRule) {
        this.rootRule = rootRule;
    }

    private CompileResult<Unit<Node>> foldEntry(Unit<String> unit) {
        System.out.println(unit.display());
        return unit.mapValue(this.rootRule::lex);
    }

    @Override
    public CompileResult<UnitSet<Node>> apply(UnitSet<String> initial) {
        return initial.iter()
                .map(this::foldEntry)
                .collect(new CompileResultCollector<>(new UnitSetCollector<>()));
    }
}
